dataset.go

```diff
@@ -86,6 +114,70 @@ func flatElementsIterator(elems []*Element, elemChan chan<- *Element) {
 	}
 }
 
+// FlatDatasetIterator represents a simple iterator over the elements in a
+// Dataset.
+type FlatDatasetIterator interface {
```

Comment:
I'd be curious about your reason for using an interface here. I only see one type that implements it: `flatIterator`, so I'm not sure what we gain from declaring an interface for it.

Interfaces require that the concrete type be resolved, so it will be a little less efficient than just returning the type outright.

Reply:
So one advantage is that the interface is exported, but the struct is not--preventing callers from creating instances of a struct that should largely be an internal mechanism. I could make the struct unexported with exported methods though, which I've seen in situations like this--but I think the Godoc is nicer with an exported interface like this.

In this case, I'm not really worried about a couple ops worth of efficiency difference. But, in this case we can probably get by with a struct with unexported internals to prevent folks from trying to make their own without the right business logic.